<a href="https://colab.research.google.com/github/stefkong1982/netology.ru/blob/Master/Rabota_s_priznakami_i_postroenie_modelej/Logisticheskaya_regressiya_SVM/DZ_Kondrstev_log_regression_SVM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнее задание к занятию «Классификация: Логистическая регрессия и SVM»

## **Задание**

**Преподаватель:** Наталья Баданина, Юлия Пономарева

**Цель:** изучить применение модели логистической регрессии и метода опорных векторов в задаче бинарной классификации.



## **Описание задания:**

В домашнем задании нужно решить задачу классификации физических лиц по уровню дохода. Данные для обучения модели хранятся в файле `adult.csv`, который можно найти в материалах к занятию или скачать [с сайта.](https://www.cs.toronto.edu/~delve/data/adult/desc.html)

* Целевая переменная – уровень дохода `income`, который принимает два значения `<=50K` и `>50K`, поэтому классификация бинарная.

* Остальные признаки описывают персональную информацию – возраст, образование, семейное положение и т. д.

* Подробное описание признаков и их возможные значения можно получить самостоятельно, используя функции `Python3` для анализа датасета (`describe`, `unique` и т.д) или прочитать информацию [по ссылке.](http://www.cs.toronto.edu/~delve/data/adult/adultDetail.html)

* Задачу классификации нужно решить при помощи обучения модели логистической регрессии и модели опорных векторов.

## **Описание Этапов работы:**

1. Получите данные и загрузите их в рабочую среду. (`Jupyter Notebook` или другую)
2. Проведите первичный анализ.
3. Проверьте данные на пропуски. Удалите в случае обнаружения. Предложите альтернативный способ работы с пропусками
4. Постройте 1-2 графика на выбор. Визуализация должна быть основана на исследуемых данных и быть полезной (из графика можно сделать вывод об особенностях датасета/класса/признака)
5. Преобразуйте категориальные признаки.
6. Разделите выборку на обучающее и тестовое подмножество. `80%` данных оставить на обучающее множество, `20%` на тестовое.
7. Обучите модели логистической регрессии и опорных векторов на обучающем множестве.
8. Для тестового множества предскажите уровень дохода и сравните с истинным значением, посчитав точность предсказания моделей. Для этого используйте встроенную функцию `score`.
9. Сформулируйте выводы по проделанной работе.
* Кратко опишите какие преобразования были сделаны с данными.
* Сравните точность двух моделей.
* Напишите свое мнение, в полной ли мере модели справились с поставленной задачей.
* Что по вашему мнению нужно сделать, чтобы улучшить результат ?




**Рекомендации к выполнению:**

* Текст оформляйте в отдельной ячейке `Jupyter Notebook/Google Colab` в формате `markdown`.
* У графиков должен быть заголовок, подписи осей, легенда (опционально).
* Делайте графики бОльшего размера, чем стандартный вывод, чтобы увеличить читаемость.
* Для получение зачета по этому домашнему заданию, минимально, должно быть реализовано обучение двух моделей классификации, выведена их точность.

## **Решение задания:**

### 1. Загрузка данных и первичный анализ:

In [27]:
import pandas as pd
# Загрузка данных
data = pd.read_csv("https://raw.githubusercontent.com/stefkong1982/netology.ru/Master/Rabota_s_priznakami_i_postroenie_modelej/Logisticheskaya_regressiya_SVM/adult.csv")

In [28]:
# Первичный анализ данных
data.head()

age  workclass  fnlwgt     education  educational-num      marital-status  \
0   25    Private  226802          11th                7       Never-married   
1   38    Private   89814       HS-grad                9  Married-civ-spouse   
2   28  Local-gov  336951    Assoc-acdm               12  Married-civ-spouse   
3   44    Private  160323  Some-college               10  Married-civ-spouse   
4   18          ?  103497  Some-college               10       Never-married   

          occupation relationship   race  gender  capital-gain  capital-loss  \
0  Machine-op-inspct    Own-child  Black    Male             0             0   
1    Farming-fishing      Husband  White    Male             0             0   
2    Protective-serv      Husband  White    Male             0             0   
3  Machine-op-inspct      Husband  Black    Male          7688             0   
4                  ?    Own-child  White  Female             0             0   

   hours-per-week native-country income  
0              40  United-States  <=50K  
1              50  United-States  <=50K  
2              40  United-States   >50K  
3              40  United-States   >50K  
4              30  United-States  <=50K

Описание атрибутов их базы данных:


Непрерывные признаки:

1. Возраст (age) - это количественное значение, представляющее возраст человека.
2. Весовой коэффициент семьи (fnlwgt) - это числовое значение, представляющее средний вес семьи для определения, насколько репрезентативна выборка в обследовании.
3. Прибыль от капитала (capital-gain) - это числовое значение, представляющее прибыль, полученную от продажи активов.
4. Потери капитала (capital-loss) - это числовое значение, представляющее убытки, понесенные от продажи активов.
5. Количество рабочих часов в неделю (hours-per-week) - это количественное значение, представляющее количество часов, которые человек работает в неделю.

Категориальные признаки:

1. Класс работы (workclass) - это категориальное значение, указывающее на тип работы, включая частный сектор, государственный сектор и самозанятых.
2. Уровень образования (education) - это категориальное значение, описывающее уровень образования, начиная от школы до докторской степени.
3. Количество лет образования (education-num) - это числовое значение, представляющее количество лет образования, завершенных человеком.
4. Семейное положение (marital-status) - это категориальное значение, указывающее на статус отношений человека, такой как женат/замужем или разведен.
5. Род занятий (occupation) - это категориальное значение, описывающее профессию или специализацию человека.
6. Отношение к респонденту (relationship) - это категориальное значение, указывающее на отношение данного человека к респонденту, например, жена, собственный ребенок и т. д.
7. Расовая принадлежность (race) - это категориальное значение, описывающее принадлежность к определенной расовой группе.
8. Пол (sex) - это категориальное значение, обозначающее пол человека как мужской или женский.
9. Страна рождения (native-country) - это категориальное значение, указывающее на страну, в которой человек родился.
10. Целевая переменная – уровень дохода (income), который принимает два значения <=50K и >50K


Более подробное описание значений признаков [см.](https://colab.research.google.com/drive/1sFzIdSsHEr-eoDqpUWNaJW56_H1ayypS?usp=sharing)

In [19]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48842 entries, 0 to 48841
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   age              48842 non-null  int64 
 1   workclass        48842 non-null  object
 2   fnlwgt           48842 non-null  int64 
 3   education        48842 non-null  object
 4   educational-num  48842 non-null  int64 
 5   marital-status   48842 non-null  object
 6   occupation       48842 non-null  object
 7   relationship     48842 non-null  object
 8   race             48842 non-null  object
 9   gender           48842 non-null  object
 10  capital-gain     48842 non-null  int64 
 11  capital-loss     48842 non-null  int64 
 12  hours-per-week   48842 non-null  int64 
 13  native-country   48842 non-null  object
 14  income           48842 non-null  object
dtypes: int64(6), object(9)
memory usage: 5.6+ MB


Общая информация о датасете:
   - Датасет содержит 48842 записи (строки) и 15 столбцов (признаков).
   - Практически все признаки имеют полное количество непропущенных значений (non-null count равен общему количеству записей для каждого столбца).
   - Информация о типах данных:
     - 6 признаков имеют тип int64 (целочисленные значения).
     - 9 признаков имеют тип object (возможно, содержат категориальные данные или строки).

In [29]:
data.describe()

age        fnlwgt  educational-num  capital-gain  \
count  48842.000000  4.884200e+04     48842.000000  48842.000000   
mean      38.643585  1.896641e+05        10.078089   1079.067626   
std       13.710510  1.056040e+05         2.570973   7452.019058   
min       17.000000  1.228500e+04         1.000000      0.000000   
25%       28.000000  1.175505e+05         9.000000      0.000000   
50%       37.000000  1.781445e+05        10.000000      0.000000   
75%       48.000000  2.376420e+05        12.000000      0.000000   
max       90.000000  1.490400e+06        16.000000  99999.000000   

       capital-loss  hours-per-week  
count  48842.000000    48842.000000  
mean      87.502314       40.422382  
std      403.004552       12.391444  
min        0.000000        1.000000  
25%        0.000000       40.000000  
50%        0.000000       40.000000  
75%        0.000000       45.000000  
max     4356.000000       99.000000

In [21]:
data.isna().sum()

age                0
workclass          0
fnlwgt             0
education          0
educational-num    0
marital-status     0
occupation         0
relationship       0
race               0
gender             0
capital-gain       0
capital-loss       0
hours-per-week     0
native-country     0
income             0
dtype: int64

Если пропуски в данных были заполнены специальным символом, например, символом "?", то такое заполнение может повлиять на работу алгоритмов машинного обучения, так как модели не смогут корректно обработать строковые значения их в таком виде.

### 2. Проверка наличия пропусков и обработка:

In [22]:
# Идентификация и замена символа "?" на пропуски (NaN):
import numpy as np
data.replace('?', np.nan, inplace=True)

In [23]:
data.isna().sum()

age                   0
workclass          2799
fnlwgt                0
education             0
educational-num       0
marital-status        0
occupation         2809
relationship          0
race                  0
gender                0
capital-gain          0
capital-loss          0
hours-per-week        0
native-country      857
income                0
dtype: int64

In [24]:
# Если есть пропуски, можно удалить строки с пропущенными значениями или заполнить их медианой или модой
data.dropna(inplace=True)

In [25]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 45222 entries, 0 to 48841
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   age              45222 non-null  int64 
 1   workclass        45222 non-null  object
 2   fnlwgt           45222 non-null  int64 
 3   education        45222 non-null  object
 4   educational-num  45222 non-null  int64 
 5   marital-status   45222 non-null  object
 6   occupation       45222 non-null  object
 7   relationship     45222 non-null  object
 8   race             45222 non-null  object
 9   gender           45222 non-null  object
 10  capital-gain     45222 non-null  int64 
 11  capital-loss     45222 non-null  int64 
 12  hours-per-week   45222 non-null  int64 
 13  native-country   45222 non-null  object
 14  income           45222 non-null  object
dtypes: int64(6), object(9)
memory usage: 5.5+ MB
